### PR TITLE
Fix: literal escape

### PR DESCRIPTION
**Pull request by liss-h**
The error was the assumption that the returned mask of `_mm256_movemask_epi8` is unsigned, which it is not.. for some reason

### CODE REVIEW

File: src/rdf4cpp/rdf/Literal.cpp [hunk 1]
```diff
@@ -36,26 +36,27 @@ static bool lexical_form_needs_escape_non_simd(std::string_view const lexical_fo
...
+bool Literal::lexical_form_needs_escape(std::string_view lexical_form) noexcept {
     // https://www.w3.org/TR/n-triples/#grammar-production-STRING_LITERAL_QUOTE
     __m256i const masks[4]{_mm256_set1_epi8('"'),
                            _mm256_set1_epi8('\\'),
                            _mm256_set1_epi8('\n'),
                            _mm256_set1_epi8('\r')};
 
-    for (size_t bix = 0; bix < lexical_form.size() / 32; ++bix) {
-        __m256i const chars = _mm256_loadu_si256(reinterpret_cast<__m256i const *>(lexical_form.data() + (bix * 32)));
+    while (lexical_form.size() >= 32) {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/Literal.cpp:46]: nice approach without index fiddeling.

File: src/rdf4cpp/rdf/Literal.cpp [hunk 2]
```diff
@@ -36,26 +36,27 @@ static bool lexical_form_needs_escape_non_simd(std::string_view const lexical_fo
...
+        __m256i const chars = _mm256_loadu_si256(reinterpret_cast<__m256i const *>(lexical_form.data()));
         for (auto const &mask : masks) {
             auto const eq = _mm256_cmpeq_epi8(mask, chars);
 
-            if (_mm256_movemask_epi8(eq) > 0) {
+            if (_mm256_movemask_epi8(eq) != 0) {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/Literal.cpp:51]: for the record: here lied the problem. negative characters were not covered

File: tests/datatype/tests_String.cpp [hunk 3]
```diff
@@ -50,6 +50,7 @@ TEST_CASE("Datatype String") {
     value = "\n";
     auto lit6 = Literal::make_typed(value, type_iri);
     CHECK(lit6.value<datatypes::xsd::String>() == value);
+    CHECK_EQ(std::string{lit6}, "\"\\n\"");
```
Comment (bigerl) [inline tests/datatype/tests_String.cpp]:
```suggestion
    CHECK_EQ(std::string{lit6}, R"("\n")");
```

File: tests/datatype/tests_String.cpp [hunk 4]
```diff
@@ -58,14 +59,17 @@ TEST_CASE("Datatype String") {
     value = "\r";
     auto lit8 = Literal::make_typed(value, type_iri);
     CHECK(lit8.value<datatypes::xsd::String>() == value);
+    CHECK_EQ(std::string{lit8}, "\"\\r\"");
```
Comment (bigerl) [inline tests/datatype/tests_String.cpp]:
```suggestion
    CHECK_EQ(std::string{lit8}, R"("\n")");
```

File: tests/datatype/tests_String.cpp [hunk 5]
```diff
@@ -58,14 +59,17 @@ TEST_CASE("Datatype String") {
...
 
     value = "\\";
     auto lit9 = Literal::make_typed(value, type_iri);
     CHECK(lit9.value<datatypes::xsd::String>() == value);
+    CHECK_EQ(std::string{lit9}, "\"\\\\\"");
```
Comment (bigerl) [inline tests/datatype/tests_String.cpp]: use also in the other changes with escape sequences raw strings

File: tests/serializer/tests_Serialize.cpp [hunk 6]
```diff
@@ -80,7 +80,8 @@ TEST_SUITE("Serialize") {
                                     "<http://url.com/subj#4> <http://url.com#pred> _:123 .\n" // blank node
                                     "<http://url.com/subj#5> <http://url.com#pred> \"Spherical Cow\"@en .\n" // rdf:langString, inlined language
                                     "<http://url.com/subj#6> <http://url.com#pred> \"Spherical Cow\"@jk .\n" // rdf:langString, non-inlined language
-                                    "<http://url.com/subj#7> <http://url.com#pred> \"Spherical Cow\"^^<http://unknow-datatype.org#dt> .\n"; // unknown datatype
+                                    "<http://url.com/subj#7> <http://url.com#pred> \"Spherical Cow\"^^<http://unknow-datatype.org#dt> .\n" // unknown datatype
+                                    "<http://url.com_(test)#8> <http://url.com#pred> \"Test \\\"Quoted\\\" String\"@en .\n";
```
Comment (bigerl) [inline tests/serializer/tests_Serialize.cpp:84]: for the whole thing raw string literals would be better.

**Review by bigerl — CHANGES_REQUESTED** [review]
Thx for the PR. Issues seems fixed and code quality improved. The readability of the tests could benefit from string literals, though.

File: tests/serializer/tests_Serialize.cpp (hunk 6 above)
Comment (liss-h) [reply tests/serializer/tests_Serialize.cpp:84]: Cannot put comments on the side if its a raw string literal

File: src/rdf4cpp/rdf/Literal.cpp (hunk 2 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/Literal.cpp:51]: *Negative masks

**Pull request merged**